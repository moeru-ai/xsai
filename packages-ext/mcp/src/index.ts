export * from './utils/get-tools'
