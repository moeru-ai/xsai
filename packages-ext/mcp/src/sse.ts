import type { SSEClientTransportOptions } from '@modelcontextprotocol/sdk/client/sse.js'

import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js'

export type CreateSSEClientsOptions = Record<string, SSEClientTransportOptions & {
  url: URL
}>

/** @experimental */
export const createSSEClients = async (mcpServers: CreateSSEClientsOptions, version: string = '1.0.0'): Promise<Record<string, Client>> =>
  Object.fromEntries(
    await Promise.all(
      Object.entries(mcpServers)
        .map(async ([serverName, { url, ...options }]) => {
          const transport = new SSEClientTransport(url, options)
          const client = new Client({ name: serverName, version })
          await client.connect(transport)

          return [serverName, client] as const
        }),
    ),
  )
