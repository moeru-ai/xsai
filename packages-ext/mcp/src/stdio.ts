import type { StdioServerParameters } from '@modelcontextprotocol/sdk/client/stdio.js'

import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js'
import { stderr } from 'node:process'

export type CreateStdioClientsOptions = Record<string, StdioServerParameters>

/** @experimental */
export const createStdioClients = async (mcpServers: CreateStdioClientsOptions, version: string = '1.0.0'): Promise<Record<string, Client>> =>
  Object.fromEntries(
    await Promise.all(
      Object.entries(mcpServers)
        .map(async ([serverName, server]) => {
          const transport = new StdioClientTransport({ ...server, stderr })
          const client = new Client({ name: serverName, version })
          await client.connect(transport)

          return [serverName, client] as const
        }),
    ),
  )
