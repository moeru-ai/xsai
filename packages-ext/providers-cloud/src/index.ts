export * from './providers'

// TODO: providers list
