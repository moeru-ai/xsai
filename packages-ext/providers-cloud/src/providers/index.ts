export { createDeepSeek } from './deepseek'
export { createFireworksAI } from './fireworks-ai'
export { createGoogleGenerativeAI } from './google-generative-ai'
export { createGroq } from './groq'
export { createMinimax } from './minimax'
export { createMistralAI } from './mistral-ai'
export { createMoonshot } from './moonshot'
export { createNovita } from './novita'
export { createOpenAI } from './openai'
export { createOpenRouter } from './openrouter'
export { createPerplexity } from './perplexity'
export { createQwen } from './qwen'
export { createSiliconFlow } from './silicon-flow'
export { createStepfun } from './stepfun'
export { createTencentHunyuan } from './tencent-hunyuan'
export { createTogetherAI } from './together-ai'
export { createWorkersAI } from './workers-ai'
export { createXAI } from './xai'
export { createZhipu } from './zhipu'
