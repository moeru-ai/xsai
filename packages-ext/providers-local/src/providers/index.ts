export { createClaude2OpenAI } from './claude2openai'
export { createOllama } from './ollama'
export { createUnSpeech } from './unspeech'
export { createUnElevenLabs } from './unspeech/elevenlabs'
export { createUnMicrosoft } from './unspeech/microsoft'
