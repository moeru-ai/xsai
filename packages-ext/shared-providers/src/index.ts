export type * from './types/create-provider-options'
export type * from './types/providers'
export * from './utils/create'
export * from './utils/merge'
