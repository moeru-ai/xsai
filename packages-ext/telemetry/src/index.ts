export { generateText } from './utils/generate-text'
export { streamText } from './utils/stream-text'
