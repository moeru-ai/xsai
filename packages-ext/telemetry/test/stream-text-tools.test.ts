import type { AttributeValue } from '@opentelemetry/api'
import type { ReadableSpan } from '@opentelemetry/sdk-trace-base'

import { InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node'
import { streamText as aiStreamText, tool as aiTool, stepCountIs } from 'ai'
import { ollama } from 'ollama-ai-provider-v2'
import { describe, expect, it } from 'vitest'
import { tool } from 'xsai'
import { z } from 'zod/v4'

import { streamText } from '../src'

describe.sequential('streamText with tools', () => {
  const memoryExporter = new InMemorySpanExporter()
  const tracerProvider = new NodeTracerProvider({
    spanProcessors: [new SimpleSpanProcessor(memoryExporter)],
  })
  tracerProvider.register()

  const trimmedAttributes = new Set<string>([
    // Not using the following because we are testing with --update
    // 'ai.response.avgOutputTokensPerSecond', // reason: time
    // 'ai.response.id', // reason: id
    // 'ai.response.msToFinish', // reason: time
    // 'ai.response.msToFirstChunk', // reason: time
    // 'ai.response.timestamp', // reason: time
    // 'ai.response.toolCalls', // reason: contains id
    // 'ai.toolCall.id', // reason: id
    // 'gen_ai.response.id', // reason: id
  ])

  const getAttributes = (span: ReadableSpan) => Object.fromEntries(
    Object.entries(span.attributes)
      .reduce<[string, AttributeValue | undefined][]>((entries, [k, v]) => {
        if (k === 'ai.response.text') {
          entries.push([k, (v as string).replaceAll('\n', ' ').slice(0, 50)])
        }
        else if (!trimmedAttributes.has(k)) {
          entries.push([k, v])
        }
        return entries
      }, []),
  )

  it('basic', async () => {
    const add = await tool({
      description: 'Adds two numbers',
      execute: ({ a, b }) => (Number.parseInt(a) + Number.parseInt(b)).toString(),
      name: 'add',
      parameters: z.object({
        a: z.string()
          .describe('First number'),
        b: z.string()
          .describe('Second number'),
      }),
    })

    let text = ''
    const { textStream } = streamText({
      baseURL: 'http://localhost:11434/v1',
      maxSteps: 5,
      messages: [{
        content: 'How many times does 114514 plus 1919810 equal? Please try to call the `add` tool to solve the problem.',
        role: 'user',
      }],
      model: 'qwen3:0.6b',
      onFinish: async () => {
        const spans = memoryExporter.getFinishedSpans().map(getAttributes)

        expect(text).toMatchSnapshot()
        expect(spans).toMatchSnapshot()
      },
      seed: 114514,
      streamOptions: {
        includeUsage: true,
      },
      tools: [add],
    })

    for await (const textDelta of textStream) {
      text += textDelta
    }
  }, 120_000)

  it('basic/ai', async () => {
    const add = aiTool({
      description: 'Adds two numbers',
      // @ts-expect-error wrong types
      execute: ({ a, b }) => (Number.parseInt(a) + Number.parseInt(b)).toString(),
      // @ts-expect-error wrong types
      inputSchema: z.object({
        a: z.string()
          .describe('First number'),
        b: z.string()
          .describe('Second number'),
      }),
      name: 'add',
    })

    const { text } = aiStreamText({
      experimental_telemetry: { isEnabled: true },
      model: ollama('qwen3:0.6b'),
      prompt: 'How many times does 114514 plus 1919810 equal? Please try to call the `add` tool to solve the problem.',
      seed: 114514,
      stopWhen: stepCountIs(5),
      tools: { add },
    })

    const fullText = await text
    const spans = memoryExporter.getFinishedSpans().slice(4).map(getAttributes)

    expect(fullText).toMatchSnapshot()
    expect(spans).toMatchSnapshot()
  }, 120_000)

  it('basic/compare', () => {
    const extractAttributeKeys = (span: ReadableSpan) => Object.keys(span.attributes)
      // eslint-disable-next-line sonarjs/no-nested-functions
      .filter(key => ['ai.settings'].every(prefix => !key.startsWith(prefix)))
      .sort((a, b) => a.localeCompare(b))

    const spans = memoryExporter.getFinishedSpans()
    const xsaiSpans = spans.slice(0, 4)
    const aiSpans = spans.slice(4)

    const xsaiIds = xsaiSpans.map(span => span.attributes['ai.operationId'])
    const aiIds = aiSpans.map(span => span.attributes['ai.operationId'])

    expect(xsaiIds).toStrictEqual(aiIds)

    const xsaiKeys = xsaiSpans.map(extractAttributeKeys)
    const aiKeys = aiSpans.map(extractAttributeKeys)

    xsaiKeys.forEach((attributes, i) => {
      expect(attributes).toStrictEqual(aiKeys[i])
    })
  })
})
