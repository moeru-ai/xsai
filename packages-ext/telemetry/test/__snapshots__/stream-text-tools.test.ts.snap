// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`streamText with tools > basic 1`] = `
"<think>
Okay, so the user is asking how many times 114514 plus 1919810 equals. Wait, that seems like adding those two numbers, right? Because the question is phrased as "how many times does 114514 plus 1919810 equal." But "equal" here might be a bit confusing. If they mean multiplying them, then obviously it's 114514 * 1919810. But maybe the user intended to ask the sum. Let me check the tools again. The only tool provided is the add function, which takes two numbers and adds them. So if the question is asking for the sum of the two numbers, I should use the add function with a and b both set to these numbers. But wait, the way it's phrased, "plus" is a bit unclear. Let me parse the question again: "How many times does 114514 plus 1919810 equal?" If we interpret "equation" as "sum," then it's simply adding 114514 and 1919810. If there's a different interpretation, but given the tools available, I should proceed with adding those two numbers. So the correct action is to call the add function with a=114514 and b=1919810, then return the result.
</think>

<think>
Okay, so the user asked how many times 114514 plus 1919810 equals. Wait, that seems like they mean the sum of those two numbers. Let me double-check. If it's just adding 114514 and 1919810, then the sum would be 114514 + 1919810.

Let me calculate that. 114514 plus 1919810. Adding them together: 114,514 + 1,919,810. Let me do that. 1,919,810 plus 114,514. Adding the digits: 1,919,810 + 0114,514. Adding those, the total should be 2,034,324 as per the tool response. So the answer is 2,034,324 times. The user probably wanted the sum, so I just need to present that result.
</think>

The sum of 114514 and 1919810 is 2,034,324."
`;

exports[`streamText with tools > basic 2`] = `
[
  {
    "ai.operationId": "ai.toolCall",
    "ai.toolCall.args": "{"a":"114514","b":"1919810"}",
    "ai.toolCall.id": "call_r9x71979",
    "ai.toolCall.name": "add",
    "ai.toolCall.result": ""2034324"",
    "operation.name": "ai.toolCall",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "xsai",
    "ai.operationId": "ai.streamText.doStream",
    "ai.prompt.messages": "[{"content":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem.","role":"user"}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"description":"Adds two numbers","inputSchema":{"$schema":"https://json-schema.org/draft/2020-12/schema","type":"object","properties":{"a":{"type":"string"},"b":{"type":"string"}},"required":["a","b"],"additionalProperties":false},"name":"add","type":"function"}",
    ],
    "ai.response.avgOutputTokensPerSecond": 53.73057459189193,
    "ai.response.finishReason": "tool_calls",
    "ai.response.id": "d0a9e4d7-fe53-4192-ae6f-94193922ab03",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.msToFinish": 6551.204834,
    "ai.response.msToFirstChunk": 3754.649459,
    "ai.response.providerMetadata": "{}",
    "ai.response.text": "<think> Okay, so the user is asking how many times",
    "ai.response.timestamp": "2025-09-03T15:53:38.599Z",
    "ai.response.toolCalls": "[{"args":"{\\"a\\":\\"114514\\",\\"b\\":\\"1919810\\"}","toolCallId":"call_r9x71979","toolCallType":"function","toolName":"add"}]",
    "ai.usage.inputTokens": 180,
    "ai.usage.outputTokens": 352,
    "ai.usage.totalTokens": 532,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "tool_calls",
    ],
    "gen_ai.response.id": "e08c033c-481a-43b3-a07e-8e5f67d2cf3c",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "xsai",
    "gen_ai.usage.input_tokens": 180,
    "gen_ai.usage.output_tokens": 352,
    "operation.name": "ai.streamText.doStream",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "xsai",
    "ai.operationId": "ai.streamText.doStream",
    "ai.prompt.messages": "[{"content":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem.","role":"user"}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"description":"Adds two numbers","inputSchema":{"$schema":"https://json-schema.org/draft/2020-12/schema","type":"object","properties":{"a":{"type":"string"},"b":{"type":"string"}},"required":["a","b"],"additionalProperties":false},"name":"add","type":"function"}",
    ],
    "ai.response.avgOutputTokensPerSecond": 68.1909050283425,
    "ai.response.finishReason": "stop",
    "ai.response.id": "a6ec18fd-ed03-431d-ab3e-0351950e65a9",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.msToFinish": 4062.1252919999997,
    "ai.response.msToFirstChunk": 1685.3098340000006,
    "ai.response.providerMetadata": "{}",
    "ai.response.text": "<think> Okay, so the user asked how many times 114",
    "ai.response.timestamp": "2025-09-03T15:53:45.151Z",
    "ai.usage.inputTokens": 517,
    "ai.usage.outputTokens": 277,
    "ai.usage.totalTokens": 794,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "428fbb0e-1092-4e47-91a4-ed4eaa800165",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "xsai",
    "gen_ai.usage.input_tokens": 517,
    "gen_ai.usage.output_tokens": 277,
    "operation.name": "ai.streamText.doStream",
  },
]
`;

exports[`streamText with tools > basic/ai 1`] = `"The sum of 114514 and 1919810 is 2034324."`;

exports[`streamText with tools > basic/ai 2`] = `
[
  {
    "ai.operationId": "ai.toolCall",
    "ai.toolCall.args": "{"a":"114514","b":"1919810"}",
    "ai.toolCall.id": "UQItbopVXKY92hiR",
    "ai.toolCall.name": "add",
    "ai.toolCall.result": ""2034324"",
    "operation.name": "ai.toolCall",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.streamText.doStream",
    "ai.prompt.messages": "[{"role":"user","content":[{"type":"text","text":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem."}]}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"type":"function","name":"add","description":"Adds two numbers","inputSchema":{"$schema":"http://json-schema.org/draft-07/schema#","type":"object","properties":{"a":{"description":"First number","type":"string"},"b":{"description":"Second number","type":"string"}},"required":["a","b"],"additionalProperties":false}}",
    ],
    "ai.response.avgOutputTokensPerSecond": 100.43444873693214,
    "ai.response.finishReason": "stop",
    "ai.response.id": "aitxt-I0HH87ni2GUz4SWw5cgXXG63",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.msToFinish": 358.44275000000016,
    "ai.response.msToFirstChunk": 346.1615419999998,
    "ai.response.providerMetadata": "{"ollama":{"responseId":null}}",
    "ai.response.text": "",
    "ai.response.timestamp": "2025-09-03T15:53:49.563Z",
    "ai.response.toolCalls": "[{"type":"tool-call","toolCallId":"UQItbopVXKY92hiR","toolName":"add","input":{"a":"114514","b":"1919810"}}]",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.inputTokens": 192,
    "ai.usage.outputTokens": 36,
    "ai.usage.totalTokens": 228,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "aitxt-I0HH87ni2GUz4SWw5cgXXG63",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "ollama.responses",
    "gen_ai.usage.input_tokens": 192,
    "gen_ai.usage.output_tokens": 36,
    "operation.name": "ai.streamText.doStream",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.streamText.doStream",
    "ai.prompt.messages": "[{"role":"user","content":[{"type":"text","text":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem."}]},{"role":"assistant","content":[{"type":"tool-call","toolCallId":"UQItbopVXKY92hiR","toolName":"add","input":{"a":"114514","b":"1919810"}}]},{"role":"tool","content":[{"type":"tool-result","toolCallId":"UQItbopVXKY92hiR","toolName":"add","output":{"type":"text","value":"2034324"}}]}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"type":"function","name":"add","description":"Adds two numbers","inputSchema":{"$schema":"http://json-schema.org/draft-07/schema#","type":"object","properties":{"a":{"description":"First number","type":"string"},"b":{"description":"Second number","type":"string"}},"required":["a","b"],"additionalProperties":false}}",
    ],
    "ai.response.avgOutputTokensPerSecond": 103.16846129408738,
    "ai.response.finishReason": "stop",
    "ai.response.id": "aitxt-9WS5jVfc3EyLoehdUnrsSEY7",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.msToFinish": 290.7865409999995,
    "ai.response.msToFirstChunk": 69.45679100000052,
    "ai.response.providerMetadata": "{"ollama":{"responseId":null}}",
    "ai.response.text": "The sum of 114514 and 1919810 is 2034324.",
    "ai.response.timestamp": "2025-09-03T15:53:49.650Z",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.inputTokens": 245,
    "ai.usage.outputTokens": 30,
    "ai.usage.totalTokens": 275,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "aitxt-9WS5jVfc3EyLoehdUnrsSEY7",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "ollama.responses",
    "gen_ai.usage.input_tokens": 245,
    "gen_ai.usage.output_tokens": 30,
    "operation.name": "ai.streamText.doStream",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.streamText",
    "ai.prompt": "{"prompt":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem."}",
    "ai.response.finishReason": "stop",
    "ai.response.providerMetadata": "{"ollama":{"responseId":null}}",
    "ai.response.text": "The sum of 114514 and 1919810 is 2034324.",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.inputTokens": 437,
    "ai.usage.outputTokens": 66,
    "ai.usage.totalTokens": 503,
    "operation.name": "ai.streamText",
  },
]
`;
